File: scripts/import-registration-cards-to-supabase.py
#!/usr/bin/env python3
import json
import os
import re
import urllib.error
import urllib.parse
import urllib.request
from pathlib import Path


ROOT_DIR = Path(__file__).resolve().parents[1]
REGISTRY_PATH = ROOT_DIR / "production" / "fichas-funcionais" / "saida" / "employee-registration-cards.json"

SUPABASE_URL = os.environ.get("VITE_SUPABASE_URL", "https://vncgdelxlevdndnyjyth.supabase.co").rstrip("/")
SUPABASE_KEY = os.environ.get("VITE_SUPABASE_PUBLISHABLE_KEY", "sb_publishable_8Cob4xt50ZC_UGD472vkrQ_IL-NUrWB")


def normalize_code(value, limit=32):
    text = str(value or "").encode("ascii", "ignore").decode("ascii").upper()
    text = re.sub(r"[^A-Z0-9]+", "_", text).strip("_")
    return (text or "SEM_CODIGO")[:limit]


def parse_date(value):
    match = re.match(r"^(\d{2})/(\d{2})/(\d{4})$", str(value or ""))
    if not match:
        return None
    return f"{match.group(3)}-{match.group(2)}-{match.group(1)}"


def parse_money(value):
    text = str(value or "").strip()
    if not text:
        return None
    try:
        return float(text.replace(".", "").replace(",", "."))
    except ValueError:
        return None


def request(method, table, query="", body=None, prefer="return=representation", timeout=15):
    url = f"{SUPABASE_URL}/rest/v1/{table}{query}"
    headers = {
        "apikey": SUPABASE_KEY,
        "Authorization": f"Bearer {SUPABASE_KEY}",
        "Content-Type": "application/json",
    }
    if prefer:
        headers["Prefer"] = prefer
    data = json.dumps(body).encode("utf-8") if body is not None else None
    req = urllib.request.Request(url, data=data, method=method, headers=headers)
    try:
        with urllib.request.urlopen(req, timeout=timeout) as response:
            raw = response.read().decode("utf-8")
            return json.loads(raw) if raw else None
    except urllib.error.HTTPError as error:
        detail = error.read().decode("utf-8")
        raise RuntimeError(f"{method} {table}{query}: {error.code} {detail}") from error


def select_one(table, filters, select="id"):
    query = "?" + urllib.parse.urlencode({"select": select, **filters})
    rows = request("GET", table, query=query, prefer=None)
    return rows[0] if rows else None


def insert_one(table, row, select="id"):
    rows = request("POST", table, body=row, prefer=f"return=representation")
    return rows[0] if rows else None


def update_one(table, record_id, row):
    query = "?" + urllib.parse.urlencode({"id": f"eq.{record_id}"})
    request("PATCH", table, query=query, body=row, prefer="return=minimal")


company_cache = {}
department_cache = {}
position_cache = {}


def get_or_create_company(card):
    code = normalize_code(card.get("company_key") or card.get("company_name"))
    if code in company_cache:
        return company_cache[code]
    existing = select_one("hr_companies", {"code": f"eq.{code}"}, "id,name")
    if existing:
        company_cache[code] = existing
        return existing
    created = insert_one("hr_companies", {
        "code": code,
        "name": card.get("company_key") or card.get("company_name") or "Sem empresa",
        "legal_name": card.get("company_name") or card.get("company_key") or "Sem empresa",
        "document_number": card.get("company_cnpj") or None,
    }, "id,name")
    company_cache[code] = created
    return created


def get_or_create_department(company_id, name):
    department_name = name or "Sem setor"
    key = f"{company_id}|{department_name}"
    if key in department_cache:
        return department_cache[key]
    existing = select_one("hr_departments", {
        "company_id": f"eq.{company_id}",
        "name": f"eq.{department_name}",
    }, "id,name")
    if existing:
        department_cache[key] = existing
        return existing
    created = insert_one("hr_departments", {
        "company_id": company_id,
        "name": department_name,
        "code": normalize_code(department_name, 20),
    }, "id,name")
    department_cache[key] = created
    return created


def get_or_create_position(card):
    name = card.get("position") or "Sem cargo"
    if name in position_cache:
        return position_cache[name]
    existing = select_one("hr_positions", {"name": f"eq.{name}"}, "id,name")
    if existing:
        position_cache[name] = existing
        return existing
    created = insert_one("hr_positions", {
        "name": name,
        "code": normalize_code(name, 24),
        "cbo_code": card.get("cbo") or None,
    }, "id,name")
    position_cache[name] = created
    return created


def find_employee(company_id, card):
    cpf = card.get("cpf")
    if cpf:
        existing = select_one("hr_employees", {"cpf": f"eq.{cpf}"}, "id,employee_code")
        if existing:
            return existing
    code = card.get("employee_code")
    if code:
        return select_one("hr_employees", {
            "company_id": f"eq.{company_id}",
            "employee_code": f"eq.{code}",
        }, "id,employee_code")
    return None


def employee_row(card, company, department, position):
    return {
        "company_id": company["id"],
        "department_id": department["id"],
        "position_id": position["id"],
        "employee_code": card.get("employee_code") or None,
        "esocial_registration": card.get("esocial_registration") or None,
        "full_name": card.get("full_name"),
        "cpf": card.get("cpf"),
        "pis_pasep": card.get("pis_pasep") or None,
        "birth_date": parse_date(card.get("birth_date")),
        "admission_date": parse_date(card.get("admission_date")),
        "termination_date": parse_date(card.get("termination_date")),
        "status": "terminated" if card.get("termination_date") else "active",
        "base_salary": parse_money(card.get("base_salary")),
        "payment_type": card.get("payment_type") or None,
        "work_schedule": (card.get("raw_sections") or {}).get("work_schedule") or None,
        "phone": card.get("phone") or None,
        "address_raw": card.get("address_raw") or None,
        "notes": "Importado da ficha funcional em PDF.",
        "raw_import": {
            "source": "registration_card_pdf",
            "card": card,
        },
    }


def import_card(card):
    company = get_or_create_company(card)
    department = get_or_create_department(company["id"], card.get("department"))
    position = get_or_create_position(card)
    row = employee_row(card, company, department, position)
    existing = find_employee(company["id"], card)
    if existing:
        update_one("hr_employees", existing["id"], row)
        return "updated"
    insert_one("hr_employees", row)
    return "created"


def main():
    cards = json.loads(REGISTRY_PATH.read_text(encoding="utf-8"))
    summary = {"created": 0, "updated": 0, "skipped": 0, "failed": 0}
    failed = []

    for index, card in enumerate(cards, 1):
        if not card.get("full_name") or not card.get("cpf"):
            summary["skipped"] += 1
            continue
        try:
            status = import_card(card)
            summary[status] += 1
            print(f"{index:03d}/{len(cards)} {status}: {card.get('company_key')} - {card.get('full_name')}")
        except Exception as error:
            summary["failed"] += 1
            failed.append({"name": card.get("full_name"), "cpf": card.get("cpf"), "error": str(error)})
            print(f"{index:03d}/{len(cards)} failed: {card.get('full_name')} - {error}")

    print(json.dumps({"total": len(cards), "summary": summary, "failed": failed}, ensure_ascii=False, indent=2))


if __name__ == "__main__":
    main()

File: scripts/import-registration-cards-to-supabase.mjs
import fs from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { createClient } from "@supabase/supabase-js";

const rootDir = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..");
const registryPath = path.join(rootDir, "production", "fichas-funcionais", "saida", "employee-registration-cards.json");

const supabaseUrl = process.env.VITE_SUPABASE_URL || "https://vncgdelxlevdndnyjyth.supabase.co";
const supabaseKey = process.env.VITE_SUPABASE_PUBLISHABLE_KEY || "sb_publishable_8Cob4xt50ZC_UGD472vkrQ_IL-NUrWB";
const supabase = createClient(supabaseUrl, supabaseKey, { auth: { persistSession: false } });

function normalizeCode(value) {
  return String(value || "")
    .normalize("NFD")
    .replace(/[\u0300-\u036f]/g, "")
    .toUpperCase()
    .replace(/[^A-Z0-9]+/g, "_")
    .replace(/^_|_$/g, "")
    .slice(0, 32) || "SEM_CODIGO";
}

function parseBrazilianDate(value) {
  const match = String(value || "").match(/^(\d{2})\/(\d{2})\/(\d{4})$/);
  if (!match) return null;
  return `${match[3]}-${match[2]}-${match[1]}`;
}

function parseBrazilianMoney(value) {
  const text = String(value || "").trim();
  if (!text) return null;
  const normalized = text.replace(/\./g, "").replace(",", ".");
  const number = Number(normalized);
  return Number.isFinite(number) ? number : null;
}

async function getOrCreateCompany(card) {
  const name = card.company_key || card.company_name || "Sem empresa";
  const code = normalizeCode(card.company_key || name);
  const existing = await supabase.from("hr_companies").select("id,name").eq("code", code).maybeSingle();
  if (existing.error) throw new Error(existing.error.message);
  if (existing.data) return existing.data;

  const created = await supabase
    .from("hr_companies")
    .insert({ code, name, legal_name: card.company_name || name, cnpj: card.company_cnpj || null })
    .select("id,name")
    .single();
  if (created.error) throw new Error(created.error.message);
  return created.data;
}

async function getOrCreateDepartment(companyId, name) {
  const departmentName = name || "Sem setor";
  const existing = await supabase
    .from("hr_departments")
    .select("id,name")
    .eq("company_id", companyId)
    .eq("name", departmentName)
    .maybeSingle();
  if (existing.error) throw new Error(existing.error.message);
  if (existing.data) return existing.data;

  const created = await supabase
    .from("hr_departments")
    .insert({ company_id: companyId, name: departmentName, code: normalizeCode(departmentName).slice(0, 20) })
    .select("id,name")
    .single();
  if (created.error) throw new Error(created.error.message);
  return created.data;
}

async function getOrCreatePosition(card) {
  const name = card.position || "Sem cargo";
  const existing = await supabase.from("hr_positions").select("id,name").eq("name", name).maybeSingle();
  if (existing.error) throw new Error(existing.error.message);
  if (existing.data) return existing.data;

  const created = await supabase
    .from("hr_positions")
    .insert({ name, code: normalizeCode(name).slice(0, 24), cbo_code: card.cbo || null })
    .select("id,name")
    .single();
  if (created.error) throw new Error(created.error.message);
  return created.data;
}

async function findExistingEmployee(companyId, card) {
  if (card.cpf) {
    const byCpf = await supabase.from("hr_employees").select("id,employee_code").eq("cpf", card.cpf).maybeSingle();
    if (byCpf.error) throw new Error(byCpf.error.message);
    if (byCpf.data) return byCpf.data;
  }

  if (card.employee_code) {
    const byCode = await supabase
      .from("hr_employees")
      .select("id,employee_code")
      .eq("company_id", companyId)
      .eq("employee_code", card.employee_code)
      .maybeSingle();
    if (byCode.error) throw new Error(byCode.error.message);
    if (byCode.data) return byCode.data;
  }

  return null;
}

function employeeRow(card, company, department, position) {
  return {
    company_id: company.id,
    department_id: department.id,
    position_id: position.id,
    employee_code: card.employee_code || null,
    esocial_registration: card.esocial_registration || null,
    full_name: card.full_name,
    cpf: card.cpf,
    pis_pasep: card.pis_pasep || null,
    birth_date: parseBrazilianDate(card.birth_date),
    admission_date: parseBrazilianDate(card.admission_date),
    termination_date: parseBrazilianDate(card.termination_date),
    status: card.termination_date ? "terminated" : "active",
    base_salary: parseBrazilianMoney(card.base_salary),
    payment_type: card.payment_type || null,
    work_schedule: card.raw_sections?.work_schedule || null,
    phone: card.phone || null,
    address_raw: card.address_raw || null,
    notes: "Importado da ficha funcional em PDF.",
    raw_import: {
      source: "registration_card_pdf",
      imported_at: new Date().toISOString(),
      card,
    },
  };
}

async function importCard(card) {
  const company = await getOrCreateCompany(card);
  const department = await getOrCreateDepartment(company.id, card.department);
  const position = await getOrCreatePosition(card);
  const existing = await findExistingEmployee(company.id, card);
  const row = employeeRow(card, company, department, position);

  if (existing) {
    const { error } = await supabase.from("hr_employees").update(row).eq("id", existing.id);
    if (error) throw new Error(error.message);
    return { status: "updated", employeeCode: row.employee_code, name: row.full_name };
  }

  const { error } = await supabase.from("hr_employees").insert(row);
  if (error) throw new Error(error.message);
  return { status: "created", employeeCode: row.employee_code, name: row.full_name };
}

const cards = JSON.parse(await fs.readFile(registryPath, "utf8"));
const results = [];

for (const card of cards) {
  if (!card.full_name || !card.cpf) {
    results.push({ status: "skipped", reason: "missing name or cpf", name: card.full_name || "" });
    continue;
  }

  try {
    results.push(await importCard(card));
  } catch (error) {
    results.push({ status: "failed", name: card.full_name, employeeCode: card.employee_code, error: error.message });
  }
}

const summary = results.reduce((acc, item) => {
  acc[item.status] = (acc[item.status] || 0) + 1;
  return acc;
}, {});

console.log(JSON.stringify({ total: results.length, summary, failed: results.filter((item) => item.status === "failed") }, null, 2));
